### PR TITLE
feat: migrations

### PR DESCRIPTION
**Pull request by richard-ramos**

### CODE REVIEW

File: nim_status/lib/migration.nim [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                          name VARCHAR NOT NULL PRIMARY KEY,
+                          hash VARCHAR NOT NULL
+                        )"""
+  db.execScript(settingSQL)
+
+
+proc getLastMigrationExecuted*(db: DbConn): MigrationResult =
+  db.createMigrationTableIfNotExists()
+  const query = "SELECT name FROM migrations ORDER BY rowid DESC LIMIT 1;"
+  for r in rows(db, query):
```
Comment (jrainville) [inline nim_status/lib/migration.nim:21]: This for seems useless since you `LIMIT 1`

**Review by jrainville — APPROVED** [review]
Nice job!

File: Makefile [hunk 2]
```diff
@@ -182,9 +182,15 @@ ifndef PCRE_LDFLAGS
  endif
 endif
 
+
+migrations:
```
[the remark this answers is not in the document]
Comment (michaelsbradleyjr) [reply Makefile:186]: Per the chat in discord just before making this second comment, I think we can have the target be the generated nim file, and `migrations` phony target can be an alias for that (see some of the other phony targets in the Makefile):

```
MIGRATIONS ?= nim_status/lib/migrations/sql_scripts.nim

$(MIGRATIONS): | deps
	$(ENV_SCRIPT) nim c -r $(NIM_PARAMS) \
		nim_status/lib/migrations/sql_generate.nim > nim_status/lib/migrations/sql_scripts.nim

migrations: $(MIGRATIONS)
```

File: Makefile [hunk 3]
```diff
@@ -182,9 +182,15 @@ ifndef PCRE_LDFLAGS
...
+	$(ENV_SCRIPT) nim c -r $(NIM_PARAMS) \
+		nim_status/lib/migrations/sql_generate.nim > nim_status/lib/migrations/sql_scripts.nim
+
+
 NIMSTATUS ?= build/nim_status.a
 
-$(NIMSTATUS): $(SQLCIPHER)
+$(NIMSTATUS): $(SQLCIPHER) migrations
```
Comment (michaelsbradleyjr) [inline Makefile:193]: In places where you have the `migrations` phony target as a dependency, instead use the `$(MIGRATIONS)` target (see my other comments).

**Review by michaelsbradleyjr — CHANGES_REQUESTED** [review]
Awesome! A few points below:

File: Makefile [hunk 4]
```diff
@@ -182,9 +182,20 @@ ifndef PCRE_LDFLAGS
...
+MIGRATIONS ?= nim_status/lib/migrations/sql_scripts.nim
+
+$(MIGRATIONS): | deps
+	$(ENV_SCRIPT) nim c -r $(NIM_PARAMS) \
+		nim_status/lib/migrations/sql_generate.nim > nim_status/lib/migrations/sql_scripts.nim
+
+clean-migration-file:
+	rm -f nim_status/lib/migrations/sql_scripts.nim
+
+migrations: clean-migration-file $(MIGRATIONS)
```
Comment (michaelsbradleyjr) [inline Makefile:194]: When introducing phony targets they should be added to the `PHONY` list at the top of the Makefile. They will work without doing that but it's best practice and avoids a situation where a file being named the same thing as a phony target causes `make` to behave unexpectedly.

**Review by michaelsbradleyjr — CHANGES_REQUESTED** [review]

File: nim_status/lib/migration.nim (hunk 1 above)
Comment (emizzle) [reply nim_status/lib/migration.nim:21]: The api has changed, it should now be similar to:
```nim
let settings = db.one(Settings, query)
if not settings.isSome:
    raise newException(ValueError, "No records found for settings")	    raise newException(ValueError, "No record found for settings")
```

File: nim_status/lib/migration.nim [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+import nimcrypto
+import chronicles
+
+type MigrationResult = Result[string, string]
+
+proc createMigrationTableIfNotExists*(db: DbConn) =
+  const settingSQL = """CREATE TABLE IF NOT EXISTS migrations (
+                          name VARCHAR NOT NULL PRIMARY KEY,
+                          hash VARCHAR NOT NULL
+                        )"""
```
Comment (emizzle) [inline nim_status/lib/migration.nim:14]:
```suggestion
  var migration: Migration
  const settingSQL = """CREATE TABLE IF NOT EXISTS {migration.tableName} (
                          {migration.name.columnName} VARCHAR NOT NULL PRIMARY KEY,
                          {migration.hash.columnName} VARCHAR NOT NULL
                        )"""
```

File: nim_status/lib/migration.nim [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+  db.execScript(settingSQL)
+
+
+proc getLastMigrationExecuted*(db: DbConn): MigrationResult =
+  db.createMigrationTableIfNotExists()
+  const query = "SELECT name FROM migrations ORDER BY rowid DESC LIMIT 1;"
```
Comment (emizzle) [inline nim_status/lib/migration.nim:20]:
```suggestion
  var migration: Migration
  const query = "SELECT {migration.name.columnName} FROM {migration.tableName} ORDER BY rowid DESC LIMIT 1;"
```

File: nim_status/lib/migration.nim [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    return MigrationResult.ok r[0].strVal
+  return MigrationResult.err "No migrations were executed"
+
+
+proc getAllMigrationsExecuted*(db: DbConn): OrderedTable[string, string] =
+  db.createMigrationTableIfNotExists()
+  const query = "SELECT name, hash FROM migrations ORDER BY rowid ASC;"
+  result = initOrderedTable[string, string]()
+  for r in rows(db, query):
+    result[r[0].strVal] = r[1].strVal
```
Comment (emizzle) [inline nim_status/lib/migration.nim:31]:
```suggestion
  var migration: Migration
  const query = "SELECT {migration.name.columnName}, {migration.hash.columnName} FROM {migration.tableName} ORDER BY rowid ASC;"
  db.all(Migration, query)
```

File: nim_status/lib/migration.nim [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+proc getLastMigrationExecuted*(db: DbConn): MigrationResult =
+  db.createMigrationTableIfNotExists()
+  const query = "SELECT name FROM migrations ORDER BY rowid DESC LIMIT 1;"
+  for r in rows(db, query):
+    return MigrationResult.ok r[0].strVal
+  return MigrationResult.err "No migrations were executed"
+
+
+proc getAllMigrationsExecuted*(db: DbConn): OrderedTable[string, string] =
```
Comment (emizzle) [inline nim_status/lib/migration.nim:26]: What do you think about returning `seq[Migration]`?
```suggestion
proc getAllMigrationsExecuted*(db: DbConn): seq[Migration] =
```

File: nim_status/lib/migration.nim [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+
+proc checkMigrations*(db: DbConn): bool =
+  let allMigrationsExecuted = db.getAllMigrationsExecuted()
+  let migrations = toSeq(migrationUp.keys)
+  
+  debug "Verifying migration data"
+
+  if toSeq(allMigrationsExecuted.keys).len > migrations.len:
+    warn "DB version might be greater than source", migrationsInCode=migrations.len, migrationsExecuted=toSeq(allMigrationsExecuted.keys).len
```
Comment (emizzle) [inline nim_status/lib/migration.nim:41]:
```suggestion
  if allMigrationsExecuted.len > migrations.len:
    warn "DB version might be greater than source", migrationsInCode=migrations.len, migrationsExecuted=allMigrationsExecuted.len
```

File: nim_status/lib/migration.nim [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    return false
+
+  var i = -1
+  for name, hash in allMigrationsExecuted.pairs:
```
Comment (emizzle) [inline nim_status/lib/migration.nim:45]:
```suggestion
  for migration in allMigrationsExecuted:
```

File: nim_status/lib/migration.nim [hunk 11]
```diff
@@ -0,0 +1,129 @@
+import sqlcipher, results
+import sequtils, tables, algorithm
+import stew/byteutils
+import migrations/sql_scripts
+import nimcrypto
+import chronicles
+
+type MigrationResult = Result[string, string]
```
Comment (emizzle) [inline nim_status/lib/migration.nim:8]:
```suggestion
type MigrationResult = Result[Migration, string]
```

File: nim_status/lib/migration.nim [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+                        )"""
+  db.execScript(settingSQL)
+
+
+proc getLastMigrationExecuted*(db: DbConn): MigrationResult =
+  db.createMigrationTableIfNotExists()
+  const query = "SELECT name FROM migrations ORDER BY rowid DESC LIMIT 1;"
+  for r in rows(db, query):
+    return MigrationResult.ok r[0].strVal
+  return MigrationResult.err "No migrations were executed"
```
Comment (emizzle) [inline nim_status/lib/migration.nim:23]:
```suggestion
  let migration = db.one(Migration, query)
  if migration.isSome:
    return MigrationResult.ok migration
  return MigrationResult.err "No migrations were executed"
```

File: nim_status/lib/migration.nim [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+      warn "Could not execute migration"
+      db.execScript("ROLLBACK")
+      return MigrationResult.err "Could not execute migration"
+
+  return db.getLastMigrationExecuted()
+
+
+proc tearDown*(db: DbConn):bool =
+  var allMigrationsExecuted = toSeq(db.getAllMigrationsExecuted().keys)
+  allMigrationsExecuted.reverse()
```
Comment (emizzle) [inline nim_status/lib/migration.nim:114]:
```suggestion
  var allMigrationsExecuted = db.getAllMigrationsExecuted().reverse()
```

File: nim_status/lib/migration.nim [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+  try:
+    db.execScript("BEGIN")
+    for migration in allMigrationsExecuted:
+      debug "Rolling back migration", migration
+      if migrationDown.hasKey(migration):
+        let script = string.fromBytes(migrationDown[migration])
+        if script != "": db.execScript(script)
+      db.exec("DELETE FROM migrations WHERE name = ?", migration)
```
Comment (emizzle) [inline nim_status/lib/migration.nim:122]: not 100% sure if this should be `migration.name` or `migration.hash`:
```suggestion
      debug "Rolling back migration", migration.name
      if migrationDown.hasKey(migration.name):
        let script = string.fromBytes(migrationDown[migration.name])
        if script != "": db.execScript(script)
      db.exec("DELETE FROM {migration.tableName} WHERE {migration.name.columnName} = ?", migration.name)
```

**Review by emizzle — CHANGES_REQUESTED** [review]

**Review by michaelsbradleyjr — APPROVED** [review]

File: nim_status/lib/accounts.nim [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+var web3_conn*: Web3
+
+proc login*(accountData, password: string) =
+  # TODO get account, validate password, etc
+  # TODO: should this be async?
+  # TODO: db should have been initialized somewhere, not here
+  # TODO: determine where will the DB connection live. In the meantime I'm storing it into a global variable
+  # TODO: determine where the web3 conn will live
+
+  let path = currentSourcePath.parentDir().parentDir().parentDir() & "/build/" & accountData
```
Comment (michaelsbradleyjr) [inline nim_status/lib/accounts.nim:15]: I forgot about this earlier, but then thought to check it again when I saw builds failing in GHA. I don't think it's causing the errors currently reported in GHA, but I still think we should probably change the path to be the current working directory until we've implemented a solution for #117.

**Review by michaelsbradleyjr — CHANGES_REQUESTED** [review]

**Review by michaelsbradleyjr — APPROVED** [review]

**Pull request merged**